nixos/modules/services/networking/fedimintd.nix

```diff
@@ -0,0 +1,363 @@
+                  prefix="''${FM_BITCOIN_RPC_URL%*@*}"  # Everything before the last '@'
+                  suffix="''${FM_BITCOIN_RPC_URL##*@}"  # Everything after the last '@'
+                  FM_BITCOIN_RPC_URL="''${prefix}:''${secret}@''${suffix}"
+                ''
+              else
+                ""
+            )
+            + ''
+              exec ${cfg.package}/bin/fedimintd
+            ''
+          );
+        in
+        {
+          description = "Fedimint Server";
+          documentation = [ "https://github.com/fedimint/fedimint/" ];
+          wantedBy = [ "multi-user.target" ];
+          environment = lib.mkMerge [
+            {
+              FM_BIND_P2P = "${cfg.p2p.bind}:${toString cfg.p2p.port}";
+              FM_BIND_API = "${cfg.api.bind}:${toString cfg.api.port}";
+              FM_P2P_URL = cfg.p2p.url;
+              FM_API_URL = cfg.api.url;
+              FM_DATA_DIR = cfg.dataDir;
+              FM_BITCOIN_NETWORK = cfg.bitcoin.network;
+              FM_BITCOIN_RPC_URL = cfg.bitcoin.rpc.url;
+              FM_BITCOIN_RPC_KIND = cfg.bitcoin.rpc.kind;
+            }
+            cfg.environment
+          ];
+          serviceConfig = {
+            User = cfg.user;
+            Group = cfg.group;
+
+            StateDirectory = "fedimintd-${fedimintdName}";
+            StateDirectoryMode = "0700";
+            ExecStart = startScript;
+
+            Restart = "always";
+            RestartSec = 10;
+            StartLimitBurst = 5;
+            UMask = "007";
+            LimitNOFILE = "100000";
+
+            LockPersonality = true;
+            MemoryDenyWriteExecute = "true";
+            NoNewPrivileges = "true";
+            PrivateDevices = "true";
+            PrivateMounts = true;
+            PrivateTmp = "true";
+            ProtectClock = true;
+            ProtectControlGroups = true;
+            ProtectHostname = true;
+            ProtectKernelLogs = true;
+            ProtectKernelModules = true;
+            ProtectKernelTunables = true;
+            ProtectSystem = "full";
+            RestrictAddressFamilies = [
+              "AF_INET"
+              "AF_INET6"
+            ];
+            RestrictNamespaces = true;
+            RestrictRealtime = true;
+            SystemCallArchitectures = "native";
+            SystemCallFilter = [
+              "@system-service"
+              "~@privileged"
+            ];
+          };
+        }
+      ))
+    ) eachFedimintd;
+
+    users.users = mapAttrs' (
+      fedimintdName: cfg:
+      (nameValuePair "fedimintd-${fedimintdName}" {
+        name = cfg.user;
+        group = cfg.group;
+        description = "Fedimint daemon user";
+        home = cfg.dataDir;
+        isSystemUser = true;
+      })
+    ) eachFedimintd;
+
+    users.groups = mapAttrs' (fedimintdName: cfg: (nameValuePair "${cfg.group}" { })) eachFedimintd;
+
+    services.nginx.virtualHosts = mapAttrs' (
+      fedimintdName: cfg:
+      (nameValuePair cfg.api.fqdn (
```

Comment:
This is a bit of an interesting trap. Further up, you allow the `fqdn` to be `null`, presumably to be able to set a conditional default. But there's a quirk in the nix language where if it's given `null` as a key in an attrset, it will silently delete that attribute.

```
nix-repl> { "hello" = 1; }
{ hello = 1; }

nix-repl> { ${null} = 1; }
{ }
```

This could mean that people who unknowingly forget to set `fqdn` suddenly doesn't find their nginx entries, despite explicitly setting `nginx.enable`.

I think the best way to fix this, is to set `fqdn`s type to `types.str` (and also `url`), not allowing null values, and setting a conditional default in `config`. That way, the module system will complain if you haven't set the value during eval. This is also more semantically correct, as `null` isn't really a valid value if you want the service to work (if I have understood correctly at least).

```nix
options.<...> = {
  fqdn = mkOption {
    type = types.str; # no null
    example = "api.myfedimint.com";
    description = "Public domain of the API address of the reverse proxy/tls terminator.";
  };
  url = mkOption {
    type = types.str; # no null
    # no default
    description = ''
      Public URL of the API address of the reverse proxy/tls terminator. Usually starting with `wss://`.
      Typically you want to override `fqdn` instead.
    '';
  };
};

config = {
  # note: no need for the url assertions anymore.

  services.fedimintd = mapAttrs (name: value: {
    api.url = lib.mkIf (value.p2p.url != null) (lib.mkDefault "fedimint://${value.p2p.fqdn}");
    p2p.url = lib.mkIf (value.api.url != null) (lib.mkDefault "wss://${value.api.fqdn}");
  }) eachFedimintd;
};
```

I'm not certain this doesn't lead to infinite recursion. It might be better to drop the magic defaults and let people do the magic themselves.